### PR TITLE
DAOS-12823 pool: fix pool upgrading unexpectedly

### PR DESCRIPTION
**Pull request by wangshilong**
1. ds_pool_mark_upgrade_completed_internal() should be only called if pool upgrading is needed.

2. let cont_iv_update() tolerate retryable errors like pool_iv_update() as this might make resumed upgrading fail.

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by wangdi1 — APPROVED** [review]

File: src/pool/srv_pool.c [hunk 1]
```diff
@@ -4905,6 +4905,7 @@ ds_pool_upgrade_if_needed(uuid_t pool_uuid, struct rsvc_hint *po_hint,
 	int				rc;
 	bool				schedule_layout_upgrade = false;
 	bool				need_put_leader = false;
+	bool				schedule_upgrade = false;
```
Comment (kccain) [inline src/pool/srv_pool.c:4908]: code readability suggestion (not a showstopper I guess):

rename schedule_upgrade -> request_schedule_upgrade (or want_schedule_upgrade)
rename schedule_layout_upgrade -> scheduled_layout_upgrade

File: src/pool/srv_pool.c [hunk 2]
```diff
@@ -4905,6 +4905,7 @@ ds_pool_upgrade_if_needed(uuid_t pool_uuid, struct rsvc_hint *po_hint,
 	int				rc;
 	bool				schedule_layout_upgrade = false;
 	bool				need_put_leader = false;
```
Comment (kccain) [inline src/pool/srv_pool.c:4907]: something to consider for code readability (at least would make it easier for me - and hopefully I have it correct below)

need_put_leader if true really means that this function is invoked as a result of the admin running the dmg command to upgrade the pool. And need_put_leader == false means it is running as part of pool service leader step up? So could the variable be named to reflect this? It seems awkward reading the various conditionals in the code and having to always remember that !need_put_leader means the function is running as part of pool step up, and that need_put_leader means the function is running as part of a dmg pool upgrade command invocation.

maybe rename need_put_leader -> dmg_upgrade_cmd ?

**Review by kccain — DISMISSED** [review]
no showstoppers, but a couple of code readability suggestions, either now or for future code updates if you agree.

**Comment by liw** [comment]
> ds_pool_mark_upgrade_completed_internal() should be only called if pool upgrading is needed.

The fix for this seems to be working:

```
[liw2@boro-66 daos]$ dmg -o ~/daos_control.yml pool query -b test-pool-0
Pool ed00a9bd-c3af-4354-a11b-820481a9e98f, ntarget=12, disabled=0, leader=1, version=1
Pool layout out of date (1 < 2) -- see `dmg pool upgrade` for details.
Pool space info:
- Disabled targets: 
- Target(VOS) count:12
- Storage tier 0 (SCM):
  Total size: 6.0 GB
  Free: 6.0 GB, min:500 MB, max:500 MB, mean:500 MB
- Storage tier 1 (NVMe):
  Total size: 24 GB
  Free: 24 GB, min:2.0 GB, max:2.0 GB, mean:2.0 GB
Rebuild idle, 0 objs, 0 recs
```

**Review by wangdi1 — APPROVED** [review]

**Pull request merged**